### ISSUE TITLE
ClassCastException occurs when converting a JSON to a record type.

### ISSUE THREAD

**Issue by ThisaruGuruge** [ballerina-platform/ballerina-lang#26632]
**Description:**
There is a `ClassCastException` throwing when trying to build a record from a JSON using the `cloneWithType()` function.
Following is the error message:

```
java.lang.ClassCastException {"message":"class io.ballerina.runtime.values.DecimalValue cannot be cast to class java.math.BigDecimal (io.ballerina.runtime.values.DecimalValue is in unnamed module of loader 'app'; java.math.BigDecimal is in module java.base of loader 'bootstrap')"}
                at ballerina.lang_value.1_0_0:fromJsonWithType(value.bal:205)
```

**Steps to reproduce:**

Define the following record types:

```ballerina
public enum ArgumentType {
    INLINE,
    VARIABLE
}

public type Scalar int|string|float|boolean;

public type Location record {|
    int line;
    int column;
|};

public type Argument record {|
    string name;
    Scalar value;
    ArgumentType 'type = INLINE;
    Location nameLocation;
    Location valueLocation;
|};

public type Field record {|
    string name;
    Argument[] arguments?;
    Field[] selections?;
    Location location;
|};

public type Operation record {|
    string name;
    OperationType 'type;
    Field[] fields?;
|};

public type Document record {|
    map<Operation> operations;
|};
```

Use the following JSON to create the `Document` record.

```JSON
{
    "operations": {
        "operation": {
            "name": "getData",
            "type": "query",
            "fields": [
                {
                    "name": "name",
                    "arguments": [
                        {
                            "name": "id",
                            "value": 132,
                            "type": "INLINE",
                            "nameLocation": {
                                "line": 8,
                                "column": 10
                            },
                            "valueLocation": {
                                "line": 8,
                                "column": 14
                            }
                        },
                        {
                            "name": "name",
                            "value": "Prof. Moriarty",
                            "type": "INLINE",
                            "nameLocation": {
                                "line": 8,
                                "column": 19
                            },
                            "valueLocation": {
                                "line": 8,
                                "column": 25
                            }
                        },
                        {
                            "name": "negative",
                            "value": -123,
                            "type": "INLINE",
                            "nameLocation": {
                                "line": 8,
                                "column": 43
                            },
                            "valueLocation": {
                                "line": 8,
                                "column": 53
                            }
                        },
                        {
                            "name": "weight",
                            "value": 75.4,
                            "type": "INLINE",
                            "nameLocation": {
                                "line": 8,
                                "column": 59
                            },
                            "valueLocation": {
                                "line": 8,
                                "column": 67
                            }
                        }
                    ],
                    "selections": [
                        {
                            "name": "first",
                            "location": {
                                "line": 9,
                                "column": 9
                            }
                        }
                    ],
                    "location": {
                        "line": 8,
                        "column": 5
                    }
                },
                {
                    "name": "id",
                    "selections": [
                        {
                            "name": "prefix",
                            "selections": [
                                {
                                    "name": "sample",
                                    "location": {
                                        "line": 14,
                                        "column": 13
                                    }
                                }
                            ],
                            "location": {
                                "line": 13,
                                "column": 9
                            }
                        },
                        {
                            "name": "suffix",
                            "location": {
                                "line": 16,
                                "column": 9
                            }
                        }
                    ],
                    "location": {
                        "line": 12,
                        "column": 5
                    }
                },
                {
                    "name": "birthdate",
                    "arguments": [
                        {
                            "name": "format",
                            "value": "DD/MM/YYYY",
                            "type": "INLINE",
                            "nameLocation": {
                                "line": 18,
                                "column": 16
                            },
                            "valueLocation": {
                                "line": 18,
                                "column": 24
                            }
                        }
                    ],
                    "location": {
                        "line": 18,
                        "column": 5
                    }
                }
            ]
        }
    }
}
```

Following is the code I used to build the record:

```ballerina
json expectedJson = readFileAndGetJson(jsonFilePath);
Document expectedDocument = check expectedJson.cloneWithType(Document);
```

**Affected Versions:**
SLP5

**OS, DB, other environment details and versions:**
MacOS

**Related Issues (optional):**
#26624

**Comment by dulvinw** [comment]
CCE is not reproducible. We will be further analyzing to identify the issue

**Comment by rdhananjaya** [comment]
Reproduced with below code on Swan Lake beta-1.

```ballerina
import ballerina/io;
import ballerina/lang.value;
public enum ArgumentType {
    INLINE,
    VARIABLE
}

public type Scalar int|string|float|boolean;

public type Location record {|
    int line;
    int column;
|};

public type Argument record {|
    string name;
    Scalar value;
    ArgumentType 'type = INLINE;
    Location nameLocation;
    Location valueLocation;
|};

public type Field record {|
    string name;
    Argument[] arguments?;
    Field[] selections?;
    Location location;
|};

public type Operation record {|
    string name;
    OperationType 'type;
    Field[] fields?;
|};

public type OperationType "query";

public type Document record {|
    map<Operation> operations;
|};

string s = "{\n" + 
                "    \"operations\": {\n" + 
                "        \"operation\": {\n" + 
                "            \"name\": \"getData\",\n" + 
                "            \"type\": \"query\",\n" + 
                "            \"fields\": [\n" + 
                "                {\n" + 
                "                    \"name\": \"name\",\n" + 
                "                    \"arguments\": [\n" + 
                "                        {\n" + 
                "                            \"name\": \"id\",\n" + 
                "                            \"value\": 132,\n" + 
                "                            \"type\": \"INLINE\",\n" + 
                "                            \"nameLocation\": {\n" + 
                "                                \"line\": 8,\n" + 
                "                                \"column\": 10\n" + 
                "                            },\n" + 
                "                            \"valueLocation\": {\n" + 
                "                                \"line\": 8,\n" + 
                "                                \"column\": 14\n" + 
                "                            }\n" + 
                "                        },\n" + 
                "                        {\n" + 
                "                            \"name\": \"name\",\n" + 
                "                            \"value\": \"Prof. Moriarty\",\n" + 
                "                            \"type\": \"INLINE\",\n" + 
                "                            \"nameLocation\": {\n" + 
                "                                \"line\": 8,\n" + 
                "                                \"column\": 19\n" + 
                "                            },\n" + 
                "                            \"valueLocation\": {\n" + 
                "                                \"line\": 8,\n" + 
                "                                \"column\": 25\n" + 
                "                            }\n" + 
                "                        },\n" + 
                "                        {\n" + 
                "                            \"name\": \"negative\",\n" + 
                "                            \"value\": -123,\n" + 
                "                            \"type\": \"INLINE\",\n" + 
                "                            \"nameLocation\": {\n" + 
                "                                \"line\": 8,\n" + 
                "                                \"column\": 43\n" + 
                "                            },\n" + 
                "                            \"valueLocation\": {\n" + 
                "                                \"line\": 8,\n" + 
                "                                \"column\": 53\n" + 
                "                            }\n" + 
                "                        },\n" + 
                "                        {\n" + 
                "                            \"name\": \"weight\",\n" + 
                "                            \"value\": 75.4,\n" + 
                "                            \"type\": \"INLINE\",\n" + 
                "                            \"nameLocation\": {\n" + 
                "                                \"line\": 8,\n" + 
                "                                \"column\": 59\n" + 
                "                            },\n" + 
                "                            \"valueLocation\": {\n" + 
                "                                \"line\": 8,\n" + 
                "                                \"column\": 67\n" + 
                "                            }\n" + 
                "                        }\n" + 
                "                    ],\n" + 
                "                    \"selections\": [\n" + 
                "                        {\n" + 
                "                            \"name\": \"first\",\n" + 
                "                            \"location\": {\n" + 
                "                                \"line\": 9,\n" + 
                "                                \"column\": 9\n" + 
                "                            }\n" + 
                "                        }\n" + 
                "                    ],\n" + 
                "                    \"location\": {\n" + 
                "                        \"line\": 8,\n" + 
                "                        \"column\": 5\n" + 
                "                    }\n" + 
                "                },\n" + 
                "                {\n" + 
                "                    \"name\": \"id\",\n" + 
                "                    \"selections\": [\n" + 
                "                        {\n" + 
                "                            \"name\": \"prefix\",\n" + 
                "                            \"selections\": [\n" + 
                "                                {\n" + 
                "                                    \"name\": \"sample\",\n" + 
                "                                    \"location\": {\n" + 
                "                                        \"line\": 14,\n" + 
                "                                        \"column\": 13\n" + 
                "                                    }\n" + 
                "                                }\n" + 
                "                            ],\n" + 
                "                            \"location\": {\n" + 
                "                                \"line\": 13,\n" + 
                "                                \"column\": 9\n" + 
                "                            }\n" + 
                "                        },\n" + 
                "                        {\n" + 
                "                            \"name\": \"suffix\",\n" + 
                "                            \"location\": {\n" + 
                "                                \"line\": 16,\n" + 
                "                                \"column\": 9\n" + 
                "                            }\n" + 
                "                        }\n" + 
                "                    ],\n" + 
                "                    \"location\": {\n" + 
                "                        \"line\": 12,\n" + 
                "                        \"column\": 5\n" + 
                "                    }\n" + 
                "                },\n" + 
                "                {\n" + 
                "                    \"name\": \"birthdate\",\n" + 
                "                    \"arguments\": [\n" + 
                "                        {\n" + 
                "                            \"name\": \"format\",\n" + 
                "                            \"value\": \"DD/MM/YYYY\",\n" + 
                "                            \"type\": \"INLINE\",\n" + 
                "                            \"nameLocation\": {\n" + 
                "                                \"line\": 18,\n" + 
                "                                \"column\": 16\n" + 
                "                            },\n" + 
                "                            \"valueLocation\": {\n" + 
                "                                \"line\": 18,\n" + 
                "                                \"column\": 24\n" + 
                "                            }\n" + 
                "                        }\n" + 
                "                    ],\n" + 
                "                    \"location\": {\n" + 
                "                        \"line\": 18,\n" + 
                "                        \"column\": 5\n" + 
                "                    }\n" + 
                "                }\n" + 
                "            ]\n" + 
                "        }\n" + 
                "    }\n" + 
                "}";

public function main() {
    json j = checkpanic value:fromJsonString(s);
    Document|error cloneWithType = j.cloneWithType(Document);
    io:println(cloneWithType);
}
```

**Comment by rdhananjaya** [comment]
This seems to be fixed with https://github.com/ballerina-platform/ballerina-lang/pull/31053.
@HindujaB please check and close this issue.

**Comment by HindujaB** [comment]
This is fixed with #31053 for beta2.
Hence, closing the issue.
